### DISCUSSION TITLE
range with units

### DISCUSSION THREAD

**Discussion by dvd101x** [josdejong/mathjs#2929] [Ideas]
Hello,

I've encountered the need to make a range with units, let's say `[1 s, 1.1 s, 1.2 s]`.

It would be nice for range to accept units like `range(1 s, 1.2 s, 0.1 s)` or `range(1 s, 1.2 s, 100 ms)`.

A work around was using `range(1, 1.2, 0.1) s`, but it would be nice for range to accept units so we could mix and match like `range(1 s, 0.02 minute, 1 ds)`.

I may be able to import the type to `range` but I was wondering if this would be desirable for this project.

**Comment by josdejong** [comment]
It is an interesting idea, thanks for sharing! I think the step size should be required in that case, there is no way to know what unit the default step of `1` should have.

**Reply by dvd101x** [reply]
An alternative would be for the step size to have the unit of the first argument like `range(30 s, 1 minute)` being equivalent to `range(30 s, 1 minute, 1 s)`.

It may be confusing, I'm just saying it might be possible.

In a similar way we can currently do `range(0, -4)` getting an empty array because the step size of 1 might not be what we intended and we need to specify it `range(0, -4, -1)`. Might be similar to doing `range(0.5 minute, 1 minute)` getting an array with only [0.5 minute] thus we would need to either change the units `range(30 s, 1 minute)` or specifying the step size `range(0.5 minute, 1 minute, 1 s)`.

**Reply by josdejong** [reply]
Yes it will be possible, but indeed I think it will be confusing. For example, what would `range(5cm + 2inch, 20cm)` do? Would it give a different result when doing `range(2inch + 5cm, 20cm)`? 

I think it would be best to simply make the step required to prevent any confusion.

**Reply by dvd101x** [reply]
You are right, I didn't think of that case.

**Comment by dvd101x** [comment]
Hi Jos, I was checking the possibility to include units in `range` and I have a few questions:

- In the factory function I noticed some dependancies with `?` like `?matrix`. What does it mean?
- If I try to include a dependance like `isPositive` I get an error:
`Error: Cannot create function "range", some dependencies are missing: "?matrix", "?bignumber", "isPositive".` I would like to know what I'm doing wrong and what does it have to do with the dependancies starting with `?`
- I noticed that the same logic is repeated on different functions to check if the iteration is still ongoing depending on the direction of `step` and if the range end is excluded. Would it be desirable if the logic was inside a function to represent that logic?

**Reply by josdejong** [reply]
1. The `?` in dependencies means the dependency is optional. In that case the implemented function must check whether the dependencies is available or not.
2. This is not related to the optional dependencies with a `?`. It means that somewhere, `createRange` is invoked without this new dependency `isPositive`. Doing a search for `createRange(` results in one place where it is hard-coded: inside `range.transform.js`. So you'll have to add the dependency there too. Though it is quite hidden, this file pops up in the stack trace of the error too.
3. If you see ways to simplify the code by extracting reused logic, I'm all in! (unless the refactored code becomes harder to understand or more complex than the original)

**Reply by dvd101x** [reply]
Thanks! everything is clear now.

My bad, I didn't notice the transform in the stack trace of the error.

**Reply by dvd101x** [reply]
I included #2997 for this
